lib/utils/docker.coffee

```diff
@@ -131,210 +123,56 @@ exports.generateConnectOpts = generateConnectOpts = (opts) ->
-						buildThroughStream = transpose.getBuildThroughStream
-							hostQemuPath: toPosixPath(qemuPath)
-							containerQemuPath: "/tmp/#{QEMU_BIN_NAME}"
-
-						newStream = stream.pipe(buildThroughStream)
-					else
-						newStream = stream
-
-					newStream
-					.pipe(logThroughStream)
-					.pipe(cacheHighlightStream())
-					.pipe(logger.streams.build)
-
-			# Create a builder
-			generateConnectOpts(options)
-			.tap (connectOpts) ->
-				ensureDockerSeemsAccessible(connectOpts)
-			.then (connectOpts) ->
-				# Allow degugging output, hidden behind an env var
-				logger.logDebug('Connecting with the following options:')
-				logger.logDebug(JSON.stringify(connectOpts, null, '  '))
-
-				builder = new dockerBuild.Builder(connectOpts)
-				opts = {}
-
-				if options.tag?
-					opts['t'] = options.tag
-				if options.nocache?
-					opts['nocache'] = true
-				if options.buildArg?
-					opts['buildargs'] = parseBuildArgs options.buildArg, (arg) ->
-						logger.logWarn("Could not parse variable: '#{arg}'")
-				if options.squash?
-					opts['squash'] = true
-
-				builder.createBuildStream(opts, hooks, reject)
-
-# Given an image id or tag, export the image to a tar archive,
-# gzip the result, and buffer it to disk.
-exports.bufferImage = (docker, imageId, bufferFile) ->
-	Promise = require('bluebird')
-	streamUtils = require('./streams')
-
-	image = docker.getImage(imageId)
-	imageMetadata = image.inspectAsync()
-
-	Promise.join image.get(), imageMetadata.get('Size'), (imageStream, imageSize) ->
-		streamUtils.buffer(imageStream, bufferFile)
-		.tap (bufferedStream) ->
-			bufferedStream.length = imageSize
+exports.generateBuildOpts = (options) ->
+	opts = {}
+	if options.tag?
+		opts['t'] = options.tag
+	if options.nocache?
+		opts['nocache'] = true
+	if options.squash?
+		opts['squash'] = true
+	if options.buildArg?
+		opts['buildargs'] = parseBuildArgs(options.buildArg)
+	return opts
 
 exports.getDocker = (options) ->
-	Docker = require('dockerode')
-	Promise = require('bluebird')
-
 	generateConnectOpts(options)
 	.tap (connectOpts) ->
 		ensureDockerSeemsAccessible(connectOpts)
-	.then (connectOpts) ->
-		# Use bluebird's promises
-		connectOpts['Promise'] = Promise
-		new Docker(connectOpts)
+	.then(createClient)
+
+exports.createClient = createClient = do ->
+	# docker-toolbelt v3 is not backwards compatible as it removes all *Async
+	# methods that are in wide use in the CLI. The workaround for now is to
+	# manually promisify the client and replace all `new Docker()` calls with
+	# this shared function that returns a promisified client.
+	#
+	# 	**New code must not use the *Async methods.**
+	#
+	Docker = require('docker-toolbelt')
+	Promise.promisifyAll Docker.prototype, {
+		filter: (name) -> name == 'run'
+		multiArgs: true
+	}
+	Promise.promisifyAll(Docker.prototype)
+	Promise.promisifyAll(new Docker({}).getImage().constructor.prototype)
+	Promise.promisifyAll(new Docker({}).getContainer().constructor.prototype)
```

Comment:
I'm not sure about docker-progress, but dockerode has promise support built-in now. Just don't pass a callback to the methods, and in the intialisation you can pass ` { Promise: require('bluebird') }` to get bluebird promises.

Reply:
I know, and I did leave a comment just above this function explaining that it's so we keep backwards compatibility with actions that use the promisified functions. New code I added only uses dockerode's methods, I just can't change and test every single use of docker outside the build/deploy commands this PR is for. The migration should be handled in separate PR and tested properly.